Step5 수동 로또 리뷰 부탁드려요

meeingjae opened the pull request:
안녕하세요! 벌써 마지막 미션이네요. 항상 세심한 리뷰 감사합니다 😊
4단계 피드백 내용도 함께 포함해서 올렸어요
날씨가 갑자기 추워졌는데요. 감기 조심하시구
좋은 주말 되세요 😊 ❤️

Commit:
refactor: step4 리뷰 사항 반영

* Score 상수 선언
* 테스트 코드 수정
* ```WinningResult``` 와 ```Score``` 관계 개선

Commit:
docs: step5 요구 사항 추가

Commit:
feat: 로또 생성 카운트 기능 구현

로또 생성 갯수와 생성 가능 갯수를 구하기 위해
```생성 카운트``` 에 대한 역할을 담당하는 ```GenerateCount``` 클래스 생성

Commit:
refactor: 당첨 결과, 로또 금액 개선

* 당첨 결과에 우승 상금을 ```Money```로 포장해 ```Money``` 클래스의 응집도 향상
* ```Money```를 생성하는 정적 팩토리 메서드 제공
* ```Money``` 클래스 사용하지 않는 메서드 제거

Commit:
refactor: 로또 번호 목록 수동 입력 기능 생성

* 유효성 검증 시점을 생성시점으로 개선

Commit:
test: 로또 번호 목록 테스트 개선

로또 번호를 생성하는 정적 팩토리 메서드로 테스트 개선

Commit:
test: 로또 목록 concat 기능 제공

수동 발급 로또 목록과 자동 발급 로또 목록을 합치기 위한 ```concat``` 메서드 제공

Commit:
test: 로또 발급기 - 수동 로또 발급 생성

Commit:
test: 로또 테스트 개선

로또 숫자 목록 클래스가 제공하는 정적 팩토리 메서드를 활용하도록 개선

Commit:
feat: 로또 수동 입력 구현

* 수동 발급 로또 수 구현
* 수동 발급 로또 번호 목록 구현
* 메서드 분리

Commit:
feat: input view 개선

수동, 자동 구매 갯수 출력 기능 구현

devyonghee commented on src/main/java/lotto/LottoBag.java:
```diff
@@ -29,7 +30,16 @@ public List<NumberBag> getLottoNumbers() {
     }
 
-    public static int availableCount(Money money) {
+    public List<Lotto> getLottoList() {
+        return this.lottoList;
+    }
```
테스트에서 비교하기 위해 추가된 메소드 같아요!!
메소드가 추가된 것은 괜찮다고 생각되긴 하지만 외부로 노출된 필드의 데이터를 보호하기 위해서 
`Collections.unmodifiableList` 으로 방어적 복사를 해주시는 방향은 어떻게 생각하시나요?

[방어적 복사란](https://velog.io/@miot2j/%EC%96%95%EC%9D%80%EB%B3%B5%EC%82%AC-%EA%B9%8A%EC%9D%80%EB%B3%B5%EC%82%AC-%EB%B0%A9%EC%96%B4%EC%A0%81-%EB%B3%B5%EC%82%AC%EB%9E%80)

또한 테스트에서 `List` 로 꺼내서 검증하는 것보다 새로운 `LottoBag` 을 생성하여 동등성(`equals` )으로 비교하는 방법도 괜찮을 것 같은데 이러한 생각에 대한 의견도 부탁드리겠습니다 🙇‍♂️

meeingjae replied:
방어적 복사의 목적 이해했습니다 ! 
```LottoBag``` 내부 속성의 주소 값이 그대로 외부로 노출 되는 것이어서, 사이드 이펙트가 발생할 수 있겠네요 !
만약 ```getLottoList``` 메서드로 로또 목록을 불러온 뒤 어떤 변경 작업이 수행되어야 한다면, 
그것은 오히려 ```LottoBag``` 내부에서 처리하는 것이 ```LottoBag```의 기능 응집도를 높히는 방향이 맞지 않을까 싶어 피드백 수용하였습니다 !
그 외에도 테스트 코드에서 객체 비교를 수행하도록 변경하였습니다 
(테스트 코드를 수정하면서 ```getLottoList``` 메서드는 사용하지 않지만, 좋은 피드백의 결과물인 것 같아 지우지 않고 남겨두었습니다 😊)
세심한 피드백 감사합니다 ! 👍👍

devyonghee commented on src/main/java/lotto/LottoBall.java:
```diff
@@ -4,8 +4,8 @@
 
-    private static final double WIN_SCORE = 1;
-    private static final double BONUS_SCORE = 0.5;
+    private static final Score WIN_SCORE = Score.of(1);
+    private static final Score BONUS_SCORE = Score.of(0.5);
```
피드백을 수용해주셔서 감사합니다 🙇‍♂️🙏

devyonghee commented on src/main/java/lotto/LottoNumberBag.java:
```diff
@@ -23,37 +27,43 @@ public Score matchScore(WinningLottoBallBag winningLottoBallBag) {
-        validNumberSize();
-        validUnique();
-        validRange();
+    private void validNumbers(List<Number> numbers) {
+        // 정적 팩토리 메서드의 객체 생성을 위해 인자값을 물고다니는 구조 괜찮은가..?
+        validNumberSize(numbers);
+        validUnique(numbers);
+        validRange(numbers);
     }
```
`validNumberSize` 와 `validUnique` 는 `LottoNumberBag` 에서 검증을 수행해주는 것은 괜찮은 것 같은데 
`numbers` 내부의 값을 `valid` 를 검증하는 것은 `LottoNumber` 내부에서만 스스로 검증해주는 방향은 어떠신가요??
자신의 값을 검증하는 부분을 외부에게 위임해주는 것 같아서요 🤔

그리고 주석에 대한 답변을 드리자면 개인적으로 어색한 구조는 아니라고 생각하는데요.
인자에 값을 주입해주는 구조이기 때문에 
메소드 시그니처만 봐도 어떤 값(인자)을 어떻게(메소드명) 검증하는지 명확하게 표현해주고 있다고 생각됩니다 😆

meeingjae replied:
피드백 주신 부분 확인하였습니다 !
```LottoNumber``` 에서 유효성 검증을 lazy하게 하는 구조였기에 그 영향을 ```LottoNumberBag``` 에도 준 것 같아요 !
피드백 수용하여 ```LottoNumber``` 에서 유효성 검증을 생성 시점에 수행하도록 개선하였고,
그 영향으로 ```LottoNumberBag```에서 ```validRange``` 작업은 제거되었습니다.
질문에 대한 꼼꼼한 답변도 감사합니다 ! 😊🙇🏻‍♂️

devyonghee commented on src/main/java/lotto/Money.java:
```diff
@@ -64,4 +42,8 @@ public boolean equals(Object o) {
         return Objects.hash(amount);
     }
+
+    public static Money of(long amount) {
+        return new Money(amount);
+    }
```
정적 팩토리 메소드를 사용해주셨네요👍
그런데 회사에서 컨벤션이 어떤지에 따라 다르겠지만,
 이전에 제가 남겨드린 자료를 참고해보시면 ([생성자 대신 정적 팩토리 메서드를 고려하라](https://sun-22.tistory.com/84)) 인자가 하나인 경우에는 `from` 을 많이 사용한다고 합니다.  
또는 `valueOf` 를 사용한다고 하는데 참고만 해주세요!!🙇‍♂️

meeingjae replied:
앗!! ```of```와 ```from``` 을 잘 구분해서 써야지! 라고 다짐했었는데
사용할 때 반대로 사용해버렸네요.. 😭😭
세심하게 봐주셔서 감사합니다 ! 이젠 햇갈리지 않을 것 같아요 !! 
피드백 감사합니다 ! 🙇🏻‍♂️

devyonghee commented on src/main/java/lotto/README.md:
````diff
@@ -109,50 +109,76 @@ chore (maintain)
 * 원시 숫자형 데이터 목록을 반환하던 ```NumberGenerator(로또 번호 생성기)```가 ```Number```의 목록을 반환하도록 개선
 * 당첨 여부 판별에 사용되던 숫자 목록을 ```당첨 공 목록```을 래핑한 클래스를 사용하는 구조로 전환
+
+## 5단계 - 로또(수동)
+
+### 추가 프로그래밍 요구사항
+
+* 모든 원시값과 문자열을 포장한다
+* 줄여쓰지 않는다 (축약 금지)
+* 모든 상황에 예외처리를 한다
+* java의 Optional을 이용해 NPE가 발생하지 않도록 한다
+
+### 추가 기능
+
+* 수동 구매 로또 수 입력
+    * 생성 카운트
+        * 속성
+            * 카운트
+        * 기능
+            * minus
+            * 카운트 값 크기 비교 유효성 검증
+* 로또 번호 생성기 (```LottoIssuer```)
+    * 수동 번호 생성
+* 수동 로또 번호 입력 (```LottoNumberBag```)
````
요구사항을 명확하게 정리해주셨네요👍👍👍👍

devyonghee commented on src/main/java/lotto/WinningResult.java:
```diff
@@ -1,40 +1,36 @@
+    WIN_THIRD(score -> score.isBetween(4, 5), Money.of(50000)),
+    WIN_SECOND(score -> score.equals(Score.of(5)), Money.of(1500000)),
+    WIN_SECOND_BONUS(score -> score.isBetween(5.5, 6), Money.of(30000000)),
+    WIN_FIRST(score -> score.isBetween(6, 7), Money.of(2000000000));
+    private Predicate<Score> matchScoreRange;
+    private final Money winningPrice;
```
`Predicate`  활용을 잘해주신 것 같아요 👍👍👍
```suggestion
    WIN_FIRST(score -> score.isBetween(6, 7), Money.of(2000000000));
    private final Predicate<Score> matchScoreRange;
    private final Money winningPrice;
```

meeingjae replied:
고민이 많았던 부분인데 감사합니다 ! 
불변 선언을 놓쳤네요 ! 
피드백 반영하였습니다 🙇🏻‍♂️

devyonghee commented on src/main/java/lotto/LottoIssuer.java:
```diff
@@ -2,21 +2,27 @@
         return new LottoBag(lottoList);
     }
 
+    public static LottoBag issueManual(List<LottoNumberBag> lottoNumberBags) {
+        return new LottoBag(lottoNumberBags.stream().map(Lotto::new)
+                .collect(Collectors.toList()));
+    }
```
`NumberGenerator` 는 로또 번호들을 생성해도록 인터페이스로 추상화를 해주신 것 같은데요. 
메뉴얼 로또를 생성하면서 활용하지 못한 점이 조금 아쉬운 것 같아요 🥲
생성해야할 로또의 갯수를 제공해서 로또 리스트를 반환해주는 동작이라면 동일하게 추상화가 가능할 것 같은데
이러한 방향에 대해서 어떻게 생각하시나요? 🤔

meeingjae replied:
피드백 주신 내용 확인하였습니다 !
초기에 로또 숫자를 생성하는 기능을 추상화해 놓은 ```Numbergenerator``` 의 수동 생성 구현체를 생성하려고 했었는데요
처음 설계할 당시엔 ```LottoNumberGenerator``` 는 정적 변수를 사용하는 유틸 성격의 클래스라고 생각해서, 속성을 가져야하는 수동 생성 구현체와는 성격이 다르지 않을까.. 라는 고민이 있었습니다.
그렇지만 피드백 주신 내용을 확인하고, 다시 곰곰히 생각해 보았을 땐
'기능 그 자체를 추상화 해 놓은 것인데, 하고자 하는 그 명확한 기능을 구현한 구현체라면, 그 것이 어떤 것이든 상관 없지 않을까'
라는 생각이 들었습니다
따라서 ```LottoNumberGenerator```의 수동 생성 구현체인 ```LottoNumberManualGenerator```를 생성하였습니다
유효성 검증에 대한 로직도 고르게 잘 분포해 있어서 바로 적용하는데 어려움이 없었습니다.
추가적인 개선 사항으로 ```수동 생성``` 기능을 위해 존재하던 ```LottoNumberBag```의 정적 팩토리 메서드도 제거되었습니다.
항상 피드백 주실 때마다 너무나도 고민해볼 가치 있는 내용들을 쏙쏙 알려주셔서
고민해보고, 개선하면서 배움과 성취감을 느낍니다
너무 감사합니다 ! 🙇🏻‍♂️ ❤️

devyonghee commented on src/main/java/lotto/LottoNumberBag.java:
```diff
@@ -1,19 +1,23 @@
 
 public class LottoNumberBag implements NumberBag {
 
+    static final String SPLIT_DELIMITER = ",";
+
```
아주아주 사소한 부분이지만 실수하신 것 같아 코멘트만 남기겠습니다 😁

```suggestion
    private static final String SPLIT_DELIMITER = ",";
```

meeingjae replied:
꼼꼼히 봐주셔서 감사합니다 !! 🙇🏻‍♂️
이런 사소한 실수를 없애고 싶은데요 😭 
꼼꼼하게 봐주신 덕에 저도 한번 더, 두번 더 확인해보는 습관을 가질 수 있을 것 같아요 !! 
감사합니다 ! 😊

devyonghee commented on src/main/java/lotto/controller/LottoStore.java:
```diff
@@ -1,32 +1,68 @@
-        LottoBag lottoList = LottoIssuer.issue(availableCount, new LottoNumberGenerator());
-        InputView.printNumbers(lottoList);
+        GenerateCount availableCount = new GenerateCount(LottoBag.availableCount(new Money(scanner.nextLine())));
+
+        System.out.println("수동으로 구매할 로또 수를 입력해 주세요.");
+        GenerateCount manualCount = new GenerateCount(scanner.nextLine());
```
문구를 노출하는 부분은 `controller` 보다 `View` 의 역할에 더 어울리 것 같은데 이러한 생각에 대해서 어떻게 생각하시나요? 🤔

meeingjae replied:
말씀하신 부분 동의합니다 !
메세지를 출력하는 부분은 ```Controller```보단 ```View```의 역할이 더 맞는 것 같아요
피드백 적극 수용하여 반영하였습니다 ! 👍

devyonghee commented on src/main/java/lotto/GenerateCount.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Objects;
+
+public class GenerateCount {
+
+    private final long count;
+
```
피드백을 수용해주시고 불변 객체를 활용해주셔서 감사합니다 🙇‍♂️

devyonghee commented on src/main/java/lotto/GenerateCount.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final long count;
+
+    public GenerateCount(String count) {
+        try {
+            this.count = Long.parseLong(count);
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException("구매 수는 숫자여야 합니다. 입력값:" + count);
+        }
+    }
+
+    public GenerateCount(long count) {
+        this.count = count;
+    }
```
필드에 값을 할당하는 부분이 두 곳으로 나뉘어진 것 같아요
코드 유지보수를 위해 주생성자에서 할당하도록 하고 부생성자에서 주생성자를 호출하는 방법은 어떠신가요? 🤔
제가 참고하고 고민했던 자료를 남겨드리도록 하겠습니다!
[생성자 하나를 주 생성자로 만드세요](https://jessyt.tistory.com/133)

meeingjae replied:
생성자 체이닝 기법에 대해서 피드백 주셨군요 !
모든 초기화는 주 생성자에서만 하는 구조로 개선하면서, 코드 내 중복 및 복잡도가 줄어들었어요
좋은 피드백 감사합니다 !! 😊👍

devyonghee reviewed (changes requested):
안녕하세요 민재님!!
드디어 마지막 단계네요👍👍🎉
피드백 반영해주시느라 정말 고생 많으셨습니다 🙇‍♂️

민재님도 감기 조심하시고 
미션하시느라 바쁘시겠지만 틈틈이 주말도 즐기셨으면 좋겠습니다 😆

역시 5단계 로또 (수동) 도 잘 구현해주신 것 같아요!!
추가로 참고하시면 더 좋을 것 같은
소소한 코멘트 몇가지 남겨드렸는데 확인 한번만 부탁드리겠습니다 🙇‍♂️

코멘트와 관련하여 궁금하거나 이상한 점이 보인다면 언제든 편하게 DM 이나 코멘트 부탁드리겠습니다. 🙏

Commit:
refactor: 로또 목록 방어적 복사 구현 및 테스트 코드 수정

* ```LottoBag.getLottoList``` 방어적 복사한 로또 목록 응답하도록 개선
* 테스트 코드에서 객체 비교를 수행하도록 개선

Commit:
refactor: 역할에 맞게 유효성 검증 로직 위임

* ```LottoNumber``` 의 유효성 검증 로직을 생성시점으로 변경
* ```LottoNumberBag``` 에서 ```LottoNumber```내부 속성 유효성 검증 하던 로직 제거

Commit:
refactor: 정적 팩토리 메서드 네이밍 컨벤션 적용

* ```of``` = 여러개의 인자를 받는 경우
* ```from``` = 한 개의 인자를 받는 경우

```Money``` 는 한개의 인자를 받아서 객체를 생성하는 정적 팩도리 메서드를 제공한다.
따라서 네이밍 컨벤션은 ```from```을 사용하도록 한다

Commit:
refactor: 내부 객체 불변 선언

```WinningResult``` 의 ```matchScoreRange``` 속성 불변 선언

Commit:
refactor: 상수 접근 지정자 지정

Commit:
refactor: 부 생성자에서 주 생성자를 호출하는 구조로 개선

Commit:
refactor: 시스템 메세지 view 영역으로 이관

Commit:
refactor: 로또 번호 수동 생성기 구현

* ```LottoNumberBag```에서 수동 생성을 위해 제공하던 정적 팩토리 메서드 제거

devyonghee commented on src/main/java/lotto/GenerateCount.java:
```diff
@@ -0,0 +1,52 @@
+        return Objects.hash(count);
+    }
+
+    public void validGreaterThan(GenerateCount manualCount) {
+        if (minus(manualCount).getCount() < 0) {
+            throw new IllegalStateException(
+                    "생성 가능한 갯수 보다 많이 입력했습니다. 생성 가능 갯수:" + count + " 수동 입력 갯수:" + manualCount.getCount());
+        }
+    }
```
적절한 예외 타입도 고민해보시면 좋을 것 같아요
예외 메세지에는 인자가 잘못되었다는 의미가 큰 것 같은데
개인적으로 `IllegalArgumentException` 이 그 의미에 더 부합할 것 같아요!

그리고 `GenerateCount` 개념은 갯수를 의미하는데 음수까지 허용해주고 있는 것 같은데요. 
갯수라는 의미에서 음수도 허용된다는 것이 괜찮을지 같이 고민해보시면 좋을 것 같아요 😆

meeingjae replied:
인자 값이 잘못 들어왔을 땐 ```IllegalArgumentException``` 참고하도록 하겠습니다 !
```GenerateCount``` 의 경우 말씀 주신 대로 예외처리에 미숙했네요 ! 
```GenerateCount``` 에서 양수만 허용하도록 유효성 검증 로직 추가하였고
비슷한 내용으로 ```Money``` 에서의 유효성 검증 로직 또한 추가하였습니다 !
피드백 감사합니다 🙇🏻‍♂️

devyonghee commented on src/test/java/lotto/LottoBagTest.java:
```diff
@@ -15,15 +18,31 @@ class LottoBagTest {
+        Lotto targetLotto = new Lotto(numberGenerator);
+        LottoBag targetBag = new LottoBag(Collections.singletonList(targetLotto));
+        //when:
+        LottoBag expected = new LottoBag(Arrays.asList(sourceLotto, targetLotto));
```
기대 값이 `when` 에 위치하는 것이 옳은 방향인지 고민해보시면 좋을 것 같아요! 😁

meeingjae replied:
```expected``` 역시 ```given```에 위치하는 것이 좋아보여요 !
어떤 행위를 나타내지 않고 검증을 위해 준비 된 데이터로서의 역할을 수행하는 변수로 보입니다
피드백 감사합니다 !! 😊👍

devyonghee commented on src/main/java/lotto/controller/LottoStore.java:
```diff
@@ -1,32 +1,69 @@
+        long count = manualCount.getCount();
+        InputView.inputManualNumbers();
+        List<LottoNumberManualGenerator> lottoNumberManualGeneratorList = new ArrayList<>();
+        for (int i = 0; i < count; i++) {
+            lottoNumberManualGeneratorList.add(new LottoNumberManualGenerator(scanner.nextLine()));
+        }
```
`for` 구문은 잘못 사용하면 의도치 않은 반복문이 될 수 있으니 
범위에 대한 동작이라면 `IntStream` 도 활용해보시는 것을 추천드릴게요!! 😆

meeingjae replied:
피드백 주신 내용 확인하였습니다 !
```LongStream``` 을 통해 ```manualCount.getCount()``` 를 소비하는 방식으로 개선하였습니다 !
소비해야 하는 데이터라면 피드백 주신 내용처럼 ```for``` 구문을 돌리는 것보다 ```Stream```을 활용하는 것도 좋다고 생각합니다 
피드백 감사합니다 ! 😊👍

devyonghee commented on src/main/java/lotto/controller/LottoStore.java:
```diff
@@ -1,32 +1,69 @@
+
+        InputView.inputManualCount();
+        GenerateCount manualCount = new GenerateCount(scanner.nextLine());
+        availableCount.validGreaterThan(manualCount);
```
더 큰 값이 들어왔을 경우 예외를 반환시키는 역할이 
어디에 위치해 있는 것이 올바른 위치인지 고민해보시면 좋을 것 같아요 🤔

meeingjae replied:
피드백 내용 확인하였습니다 !
```minus``` 메서드 내부 로직으로 옮겨 ```GenerateCount``` 클래스의 응집도를 높혔습니다
유효성 검증이 ```GenerateCount``` 클래스 내부에서 이루어 질 수 있었는데 외부로 노출되어 버렸네요..
세심하게 봐주셔서 감사합니다! 피드백 반영하였습니다
감사합니다 🙇🏻‍♂️ ❤️

devyonghee reviewed (approved):
민재님 마지막 단계까지 피드백 반영해주시느라 고생 많으셨습니다 🙇‍♂️🙇‍♂️🙇‍♂️
군데군데 고민해주신 부분들이 많이 보이는 것 같아요 👍👍

열심히 하려고 했는데 리뷰어로써 많이 부족하지는 않았는지 걱정이네요 🥲

추가적으로 한번 더 고민해보시면 좋을 것 같은 부분들 코멘트 남겨드렸습니다!
로또 미션 진행하시느라 정말 고생 많으셨고
남은 미션도 열심히 하셔서 꼭 좋은 결과 있으셨으면 좋겠습니다 😆👍
